internal/http/services/owncloud/ocdav/ocdav.go

```diff
@@ -235,20 +236,59 @@ func (s *svc) getClient() (gateway.GatewayAPIClient, error) {
 	return pool.GetGatewayServiceClient(s.c.GatewaySvc)
 }
 
-func applyLayout(ctx context.Context, ns string, useLoggedInUserNS bool, requestPath string) string {
+func (s *svc) ApplyLayout(ctx context.Context, ns string, useLoggedInUserNS bool, requestPath string) (string, string, error) {
 	// If useLoggedInUserNS is false, that implies that the request is coming from
 	// the FilesHandler method invoked by a /dav/files/fileOwner where fileOwner
 	// is not the same as the logged in user. In that case, we'll treat fileOwner
 	// as the username whose files are to be accessed and use that in the
 	// namespace template.
 	u, ok := ctxpkg.ContextGetUser(ctx)
 	if !ok || !useLoggedInUserNS {
-		requestUserID, _ := router.ShiftPath(requestPath)
-		u = &userpb.User{
-			Username: requestUserID,
+		// take a shortcut if "ns" is not a template
+		if !strings.Contains(ns, "{{") {
+			return ns, requestPath, nil
 		}
+		var requestUsernameOrID string
+		requestUsernameOrID, requestPath = router.ShiftPath(requestPath)
+
+		gatewayClient, err := s.getClient()
+		if err != nil {
+			return "", "", err
+		}
+
+		// Check if we got a Username
+		res, err := gatewayClient.GetUserByClaim(ctx, &userpb.GetUserByClaimRequest{
```

Comment:
These calls can be avoided by checking what is being specified in the template. If we know that the calls will always be to the `/dav/files/$username/$path` and the namespace template asks for username only, then we can return early. Also, these need to be cached. Going to the userprovider service for every storage operation would be costly.